core/drivers/crypto/hisilicon/sec_hash.c

```diff
@@ -0,0 +1,674 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022-2024 HiSilicon Limited.
+ * Kunpeng hardware accelerator sec hash algorithm implementation.
+ */
+
+#include <drvcrypt_hash.h>
+#include <initcall.h>
+#include <sec_hash.h>
+#include <sec_main.h>
+
+static enum hisi_drv_status sec_digest_set_hmac_key(struct hashctx *ctx,
+						    struct hisi_sec_sqe *sqe)
+{
+	if (ctx->key_len > SEC_DIGEST_MAX_KEY_SIZE || !ctx->key_len) {
+		EMSG("Invalid digest key len(%ld)", ctx->key_len);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+
+	/* If the length of key is not word-aligned, increment by 1 */
+	sqe->type2.mac_key_alg |= ((ctx->key_len + SEC_ENCODE_BYTES - 1) /
+				   SEC_ENCODE_BYTES) << SEC_AKEY_OFFSET;
+	sqe->type2.a_key_addr = ctx->key_dma;
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void sec_digest_fill_long_bd2(struct hashctx *ctx,
+				     struct hisi_sec_sqe *sqe)
+{
+	uint64_t total_bits = 0;
+
+	if (ctx->has_next && !ctx->iv_len) {
+		/* LONG BD FIRST */
+		sqe->ai_apd_cs |= AI_GEN_INNER;
+		sqe->ai_apd_cs |= (AUTHPAD_NOPAD << SEC_APAD_OFFSET);
```

Comment:
`SHIFT_U32()` or `SHIFT_U64()` are generally preferred since they are safer.

Reply:
ai_apd_cs is u8

Reply:
Sure in this particular case there no issue since `SEC_APAD_OFFSET` is `2`, but you you need to crosscheck to see that it isn't 31 in case `AUTHPAD_NOPAD` is a signed 32-bit integer. So it's better to use these shift macros where possible.